c/str/split_into_nhot.cc

```diff
@@ -0,0 +1,159 @@
+//------------------------------------------------------------------------------
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// © H2O.ai 2018
+//------------------------------------------------------------------------------
+#include "str/py_str.h"
+#include <unordered_map>
+#include <vector>
+#include "datatable.h"
+#include "utils/exceptions.h"
+#include "utils/omp.h"
+#include "utils/shared_mutex.h"
+
+namespace dt {
+
+/**
+ * Split string into tokens
+ */
+static void tokenize_string(
+    std::vector<std::string>& tokens,
+    const char* strstart,
+    const char* strend,
+    char sep
+) {
+  tokens.clear();
+  const char* ch = strstart;
+  while (ch < strend) {
+    if (*ch == ' ' || *ch == '\t' || *ch == '\n' || *ch == sep) {
+      ch++;
+    }
+    else {
+      const char* ch0 = ch;
+      // Handle the case of quoted token
+      if (*ch == '\'' || *ch == '"') {
+        char quote = *ch++;
+        while (ch < strend && *ch != quote) {
+          ch += 1 + (*ch == '\\');
+        }
+        if (ch < strend) {
+          tokens.emplace_back(ch0 + 1, ch);
+          ch++;  // move over the closing quote
+          continue;
+        }
+        // fall-through
+        ch = ch0;
+      }
+      // Regular non-quoted token, parse until the next sep
+      while (ch < strend && *ch != sep) {
+        ch++;
+      }
+      const char* ch1 = ch;
+      while (ch1 > ch0) {  // also omit trailing spaces
+        char c = *(ch1 - 1);
+        if (!(c == ' ' || c == '\t' || c == '\n')) break;
+        ch1--;
+      }
+      tokens.emplace_back(ch0, ch1);
+      ch++;  // move over the sep
+    }
+  }
+}
```

Comment:
Is there any reason not to use `strtok()` function from `<cstring>`? Performance?

Reply:
It doesn't have the ability to handle quotes, it doesn't strip whitespace, and it modifies the string that is passed to it, so I'm not even sure whether it's safe to use with `std::string`.

Reply:
I see, yes, trimming and quotes should be done separately. I guess one can't use it with `std::string` directly, so a duplicate created with something like `strdup()` may be needed.